### PR TITLE
created User entity and service for it, AuthenticationController,

### PR DESCRIPTION
**Pull request by OlgaStarikova**
Fieldmatch annotation

### CODE REVIEW

File: src/main/java/com/example/onlinebookstore/dto/UserResponseDto.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package com.example.onlinebookstore.dto;
+
+import lombok.Data;
+
+@Data
+public class UserResponseDto {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/dto/UserResponseDto.java:6]: Let's use record, it perfectly fits to DTO

File: src/main/java/com/example/onlinebookstore/dto/UserRegistrationRequestDto.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+import com.example.onlinebookstore.validation.FieldMatch;
+import jakarta.validation.constraints.Email;
+import jakarta.validation.constraints.NotBlank;
+import lombok.Data;
+import org.hibernate.validator.constraints.Length;
+
+@Data
+@FieldMatch(first = "password", second = "repeatPassword")
+public class UserRegistrationRequestDto {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/dto/UserRegistrationRequestDto.java:11]: Let's use record, it perfectly fits to DTO

File: src/main/java/com/example/onlinebookstore/exception/RegisterationException.java [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package com.example.onlinebookstore.exception;
+
+public class RegisterationException extends RuntimeException {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/exception/RegisterationException.java:3]:
```suggestion
public class RegistrationException extends RuntimeException {
```

File: src/main/java/com/example/onlinebookstore/model/User.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+import jakarta.persistence.Column;
+import jakarta.persistence.Entity;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import lombok.Data;
+
+@Entity
+@Data
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/model/User.java:11]: Don't use @Data for entities
```suggestion
@Table(.....
```

File: src/main/java/com/example/onlinebookstore/repository/user/UserRepository.java [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package com.example.onlinebookstore.repository.user;
+
+import com.example.onlinebookstore.model.User;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface UserRepository extends JpaRepository<User, Long>, JpaSpecificationExecutor<User> {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/repository/user/UserRepository.java:9]: Why do you need JpaSpecificationExecutor<User> ?

File: src/main/java/com/example/onlinebookstore/repository/user/UserRepository.java [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+    User findByEmail(String email);
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/repository/user/UserRepository.java:10]:
```suggestion
    Optional<User> findByEmail(String email);
```

File: src/main/java/com/example/onlinebookstore/service/impl/UserServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto requestDto) {
+        if (userRepository.findByEmail(requestDto.getEmail()) != null) {
+            throw new RegisterationException("The user with email "
+                    + requestDto.getEmail()
+                    + "is already registered");
+        }
+        User user = new User();
+        user.setEmail(requestDto.getEmail());
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/service/impl/UserServiceImpl.java:27]: this should be done by mapper

File: src/main/java/com/example/onlinebookstore/service/impl/UserServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+@RequiredArgsConstructor
+@Service
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final UserMapper userMapper;
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto requestDto) {
+        if (userRepository.findByEmail(requestDto.getEmail()) != null) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/service/impl/UserServiceImpl.java:21]: It's better to create another method in user repo
```suggestion
        if (userRepo.existsByEmail(request.email())) {
```

File: src/main/java/com/example/onlinebookstore/validation/FieldMatch.java [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Constraint(validatedBy = FieldMatchValidator.class)
+@Target({ElementType.TYPE, ElementType.PARAMETER, ElementType.FIELD})
+@Retention(RetentionPolicy.RUNTIME)
+public @interface
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/validation/FieldMatch.java:14]:
```suggestion
public @interface FieldMatch {
```

File: src/main/java/com/example/onlinebookstore/validation/FieldMatch.java [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+    Class<? extends Payload>[] payload() default {};
+
+    String first();
+
+    String second();
+
+    String message() default "fields aren't matched";
+
+
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/validation/FieldMatch.java:26]:
```suggestion
```

File: src/main/java/com/example/onlinebookstore/validation/FieldMatchValidator.java [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+    private String firstFieldName;
+    private String secondFieldName;
+    private String message;
+
+    @Override
+    public void initialize(FieldMatch constraintAnnotation) {
+        firstFieldName = constraintAnnotation.first();
+        secondFieldName = constraintAnnotation.second();
+        message = constraintAnnotation.message();
+
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/validation/FieldMatchValidator.java:17]:
```suggestion
```

File: src/main/java/com/example/onlinebookstore/validation/FieldMatchValidator.java [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    @Override
+    public boolean isValid(Object value, ConstraintValidatorContext context) {
+        boolean valid = true;
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/validation/FieldMatchValidator.java:22]:
```suggestion
        Object field = new BeanWrapperImpl(value).getPropertyValue(this.field);
        Object fieldMatch = new BeanWrapperImpl(value).getPropertyValue(this.fieldMatch);
        return Objects.equals(field, fieldMatch);
```

File: pom.xml [hunk 13]
```diff
@@ -82,6 +82,12 @@
             <artifactId>hibernate-validator</artifactId>
         </dependency>
 
+        <dependency>
+            <groupId>commons-beanutils</groupId>
```
Comment (Elena-Bruyako) [inline pom.xml:86]: You actually don't need this dependency. See comment in FieldMatchValidator

File: src/main/java/com/example/onlinebookstore/validation/Isbn.java [hunk 14]
```diff
@@ -10,7 +10,8 @@
 @Constraint(validatedBy = IsbnValidator.class)
 @Target({ElementType.PARAMETER, ElementType.FIELD})
 @Retention(RetentionPolicy.RUNTIME)
-public @interface Isbn {
+public @interface
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/validation/Isbn.java:13]: Do you use this annotation somewhere?

File: src/main/resources/db/changelog/changes/02-create-user-table.yml [hunk 15]
```diff
@@ -0,0 +1,41 @@
+databaseChangeLog:
+  - changeSet:
+      id: create-user-table
+      author: Olya Starykova
+      changes:
+        - createTable:
+            tableName: user
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/02-create-user-table.yml:7]: Table name should be plural
```suggestion
            tableName: users
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution!

File: src/main/java/com/example/onlinebookstore/exception/CustomGlobalExceptionHandler.java [hunk 16]
```diff
@@ -63,4 +63,13 @@ protected ResponseEntity<Object> handleSqlIntegrityConstraintViolationException(
...
         return new ResponseEntity<>(responseBody, HttpStatus.BAD_REQUEST);
     }
+
+    @ExceptionHandler({RegistrationException.class})
+    protected ResponseEntity<Object> handleRegistrationException(RegistrationException ex) {
+        Map<String, Object> responseBody = new LinkedHashMap<>();
+        responseBody.put("timestamp", LocalDateTime.now());
+        responseBody.put("status", HttpStatus.BAD_REQUEST);
+        responseBody.put("error", ex.getMessage());
+        return new ResponseEntity<>(responseBody, HttpStatus.BAD_REQUEST);
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/exception/CustomGlobalExceptionHandler.java:73]:
```suggestion
        return new ResponseEntity<>(responseBody, CONFLICT);
```

File: src/main/java/com/example/onlinebookstore/model/Book.java [hunk 17]
```diff
@@ -5,15 +5,17 @@
...
 import jakarta.persistence.GenerationType;
 import jakarta.persistence.Id;
+import jakarta.persistence.Table;
 import java.math.BigDecimal;
 import lombok.Data;
 import org.hibernate.annotations.SQLDelete;
 import org.hibernate.annotations.SQLRestriction;
 
 @Entity
 @Data
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/model/Book.java:15]: [Read an article](https://medium.com/@miguelangelperezdiaz444/the-hidden-dangers-of-lombok-annotations-in-your-java-code-what-you-need-to-know-8acdce2d6b89) about Data annotation and why you shouldn't use it

File: src/main/java/com/example/onlinebookstore/model/Book.java [hunk 18]
```diff
@@ -5,15 +5,17 @@
...
-@SQLDelete(sql = "UPDATE book SET is_deleted = true WHERE id = ?")
+@SQLDelete(sql = "UPDATE books SET is_deleted = true WHERE id = ?")
 @SQLRestriction("is_deleted = false")
+@Table(name = "books")
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/model/Book.java:18]: You should add constraints for required fields

File: src/main/java/com/example/onlinebookstore/model/User.java (hunk 4 above)
Comment (Elena-Bruyako) [reply src/main/java/com/example/onlinebookstore/model/User.java:12]: See comment above

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Let's fix some comments!

File: src/main/java/com/example/onlinebookstore/model/Book.java (hunk 17 above)
Comment (OlgaStarikova) [reply src/main/java/com/example/onlinebookstore/model/Book.java:15]: thank you!

File: src/main/resources/db/changelog/changes/02-create-user-table.yml [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+                    nullable: false
+              - column:
+                  name: email
+                  type: varchar(254)
+                  constraints:
+                    nullable: false
+                    unique: true
+              - column:
+                  name: password
+                  type: varchar(254)
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/02-create-user-table.yml:24]: The size should be the same as in you request. Check everywhere

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution!

File: src/main/java/com/example/onlinebookstore/model/Book.java [hunk 20]
```diff
@@ -5,26 +5,93 @@
...
 public class Book {
     @Id
     @GeneratedValue(strategy = GenerationType.IDENTITY)
     private Long id;
+    @Column(nullable = false)
     private String title;
+    @Column(nullable = false)
     private String author;
     @Column(unique = true)
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/model/Book.java:25]:
```suggestion
    @Column(unique = true, nullable = false)
```

File: src/main/java/com/example/onlinebookstore/model/Book.java [hunk 21]
```diff
@@ -5,26 +5,93 @@
...
+import jakarta.persistence.Table;
 import java.math.BigDecimal;
-import lombok.Data;
 import org.hibernate.annotations.SQLDelete;
 import org.hibernate.annotations.SQLRestriction;
 
 @Entity
-@Data
-@SQLDelete(sql = "UPDATE book SET is_deleted = true WHERE id = ?")
+@SQLDelete(sql = "UPDATE books SET is_deleted = true WHERE id = ?")
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/model/Book.java:14]:
```suggestion
@Getter
@Setter
@SQLDelete(sql = "UPDATE books SET is_deleted = true WHERE id = ?")
```

File: src/main/java/com/example/onlinebookstore/model/Book.java [hunk 22]
```diff
@@ -5,26 +5,93 @@
...
     @Column(unique = true)
     private String isbn;
+    @Column(nullable = false)
     private BigDecimal price;
     private String description;
     private String coverImage;
     @Column(nullable = false)
     private boolean isDeleted = false;
+
+    public Long getId() {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/model/Book.java:34]: remove all get and set methods, use annotations

File: src/main/java/com/example/onlinebookstore/model/User.java [hunk 23]
```diff
@@ -0,0 +1,74 @@
+package com.example.onlinebookstore.model;
+
+import jakarta.persistence.Column;
+import jakarta.persistence.Entity;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+
+@Entity
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/model/User.java:10]:
```suggestion
@Entity
@Getter
@Setter
```

Do we need soft delete approach here?

File: src/main/java/com/example/onlinebookstore/repository/user/UserRepository.java [hunk 24]
```diff
@@ -0,0 +1,13 @@
...
+import com.example.onlinebookstore.model.User;
+import java.util.Optional;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface UserRepository extends JpaRepository<User, Long> {
+    Optional<User> findByEmail(String email);
+
+    Boolean existsByEmail(String email);
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/repository/user/UserRepository.java:12]:
```suggestion
    boolean existsByEmail(String email);
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Let's fix some comments.

File: src/main/java/com/example/onlinebookstore/model/User.java [hunk 25]
```diff
@@ -0,0 +1,84 @@
...
+    private String password;
+    @Column(nullable = false)
+    private String firstName;
+    @Column(nullable = false)
+    private String lastName;
+    private String shippingAddress;
+    @Column(nullable = false)
+    private boolean isDeleted = false;
+
+    public Long getId() {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/model/User.java:36]: Be attentive, you added Getter and Setter annotation, so you don't need these methods

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
One fix needed

**Review by Elena-Bruyako — APPROVED** [review]
Well done!

**Pull request merged**